hw/xbox/nv2a/nv2a_pgraph.c

```diff
@@ -299,26 +299,81 @@ static uint64_t fnv_hash(const uint8_t *data, size_t len);
 static uint64_t fast_hash(const uint8_t *data, size_t len, unsigned int samples);
 
 /* PGRAPH - accelerated 2d/3d drawing engine */
+
+static uint32_t pgraph_rdi_read(PGRAPHState *pg,
+                                unsigned int select, unsigned int address)
+{
+    uint32_t r = 0;
+    switch(select) {
+    case RDI_INDEX_VTX_CONSTANTS0:
+    case RDI_INDEX_VTX_CONSTANTS1:
+        assert((address / 4) < NV2A_VERTEXSHADER_CONSTANTS);
+        r = pg->vsh_constants[address / 4][3 - address % 4];
+        break;
+    default:
+        fprintf(stderr, "nv2a: unknown rdi read select 0x%x address 0x%x\n", select, address);
+        assert(false);
+        break;
+    }
+    return r;
+}
+
+static void pgraph_rdi_write(PGRAPHState *pg,
+                             unsigned int select, unsigned int address,
+                             uint32_t val)
+{
+    switch(select) {
+    case RDI_INDEX_VTX_CONSTANTS0:
+    case RDI_INDEX_VTX_CONSTANTS1:
+        assert(false); /* Untested */
+        assert((address / 4) < NV2A_VERTEXSHADER_CONSTANTS);
+        pg->vsh_constants_dirty[address / 4] |=
+            (val != pg->vsh_constants[address / 4][3 - address % 4]);
+        pg->vsh_constants[address / 4][3 - address % 4] = val;
+        break;
+    default:
+        fprintf(stderr, "nv2a: unknown rdi write select 0x%x, address 0x%x\n", select, address);
+        assert(false);
```

Comment:
Should we drop the assert here? From your PR comments:
"The fact that D3D uses this prevents us from `assert(false)` on writes."
"After merge, I'll create issues about other `select` values which will be required for this feature to be useful. Issues will only focus on those features we need, not completeness."

You seem to suggest we don't want to assert here, should we remove `assert(false)` for non-implemented rdi write selects?

Reply:
Good catch! I had fixes that I forgot to push.